### PR TITLE
EVG-3167 Return some generate.tasks errors immediately

### PR DESCRIPTION
**Pull request by bsamek**

### CODE REVIEW

File: rest/data/generate.go [hunk 1]
```diff
@@ -13,29 +14,29 @@ import (
 type GenerateConnector struct{}
 
 // GenerateTasks parses JSON files for `generate.tasks` and creates the new builds and tasks.
-func (gc *GenerateConnector) GenerateTasks(taskID string, jsonBytes []json.RawMessage) error {
+func (gc *GenerateConnector) GenerateTasks(taskID string, jsonBytes []json.RawMessage) (int, error) {
```
Comment (richardsamuels) [inline rest/data/generate.go]: Why not just return rest.APIError's from this function

File: rest/client/request.go [hunk 2]
```diff
@@ -198,6 +200,12 @@ func (c *communicatorImpl) retryRequest(ctx context.Context, info requestInfo, d
 				return resp, nil
 			} else if resp.StatusCode == http.StatusConflict {
 				return nil, HTTPConflictError
+			} else if resp.StatusCode == http.StatusBadRequest {
```
Comment (richardsamuels) [inline rest/client/request.go]: all 400 status codes are considered unrecoverable, so let's just return for any 4xx code.

Also this is probably a good time to mention rehttp, the library with an http client that deals with retry logic, and we can replace this code with it?

File: model/generate.go [hunk 3]
```diff
@@ -99,38 +101,38 @@ func ParseProjectFromJSON(data []byte) (GeneratedProject, error) {
 
 // NewVersion adds the buildvariants, tasks, and functions
 // from a generated project config to a project.
-func (g *GeneratedProject) NewVersion() (*Project, *version.Version, *task.Task, *projectMaps, error) {
+func (g *GeneratedProject) NewVersion() (*Project, *version.Version, *task.Task, *projectMaps, int, error) {
```
Comment (richardsamuels) [inline model/generate.go]: Same thing here, just return rest.APIErrors?

File: rest/data/generate.go (hunk 1 above)
Comment (tychoish) [reply rest/data/generate.go]: I like that idea better than returning HTTP codes from the model/data access layer. 

If you need more than on error I'd use common error values with error.Cause()

File: model/generate.go [hunk 4]
```diff
@@ -103,32 +106,32 @@ func (g *GeneratedProject) NewVersion() (*Project, *version.Version, *task.Task,
 	// Get task, version, and project.
 	t, err := task.FindOneId(g.TaskID)
 	if err != nil {
-		return nil, nil, nil, nil, errors.Wrapf(err, "error finding task %s", g.TaskID)
+		return nil, nil, nil, nil, &rest.APIError{StatusCode: http.StatusInternalServerError, Message: errors.Wrapf(err, "error finding task %s", g.TaskID).Error()}
 	}
```
Comment (richardsamuels) [inline model/generate.go]: nil check on t, return BadRequest if nil

File: model/generate.go [hunk 5]
```diff
@@ -103,32 +106,32 @@ func (g *GeneratedProject) NewVersion() (*Project, *version.Version, *task.Task,
...
 	v, err := version.FindOneId(t.Version)
 	if err != nil {
-		return nil, nil, nil, nil, errors.Wrapf(err, "error finding version %s", t.Version)
+		return nil, nil, nil, nil, &rest.APIError{StatusCode: http.StatusInternalServerError, Message: errors.Wrapf(err, "error finding version %s", t.Version).Error()}
 	}
```
Comment (richardsamuels) [inline model/generate.go]: nil check on v, return BadRequest if nil

File: rest/client/request.go [hunk 6]
```diff
@@ -198,6 +200,12 @@ func (c *communicatorImpl) retryRequest(ctx context.Context, info requestInfo, d
 				return resp, nil
 			} else if resp.StatusCode == http.StatusConflict {
 				return nil, HTTPConflictError
+			} else if resp.StatusCode >= 400 && resp.StatusCode < 500 {
+				apiErr := rest.APIError{}
+				if err := util.ReadJSONInto(resp.Body, err); err != nil {
+					return nil, errors.Wrapf(apiErr, "server returned %d", http.StatusBadRequest)
```
Comment (richardsamuels) [inline rest/client/request.go]: Wrapf(&apiError), because errors are interfaces. (or change `apiError := rest...` to `apiError := &rest...`)

**Pull request merged**